kernel/src/sched/round_robin.rs

```diff
@@ -0,0 +1,203 @@
+//! Round Robin Scheduler for Tock
+//! This scheduler is specifically a Round Robin Scheduler with Interrupts.
+//! When hardware interrupts occur while a userspace process is executing,
+//! this scheduler executes the top half of the interrupt,
+//! and then stops executing the userspace process immediately and handles the bottom
+//! half of the interrupt. This design decision was made to mimic the behavior of the
+//! original Tock scheduler. In order to ensure fair use of timeslices, when
+//! userspace processes are interrupted the systick is paused, and the same process
+//! is resumed with the same systick value from when it was interrupted.
+
+use crate::callback::AppId;
+use crate::capabilities;
+use crate::common::dynamic_deferred_call::DynamicDeferredCall;
+use crate::common::list::{List, ListLink, ListNode};
+use crate::ipc;
+use crate::platform::mpu::MPU;
+use crate::platform::systick::SysTick;
+use crate::platform::{Chip, Platform};
+use crate::process;
+use crate::sched::{Kernel, Scheduler};
+use crate::syscall::{ContextSwitchReason, Syscall};
+
+/// A node in the linked list the scheduler uses to track processes
+pub struct RRProcessNode<'a> {
+    appid: AppId,
+    next: ListLink<'a, RRProcessNode<'a>>,
+}
+
+impl<'a> RRProcessNode<'a> {
+    pub fn new(appid: AppId) -> RRProcessNode<'a> {
+        RRProcessNode {
+            appid: appid,
+            next: ListLink::empty(),
+        }
+    }
+}
+
+impl<'a> ListNode<'a, RRProcessNode<'a>> for RRProcessNode<'a> {
+    fn next(&'a self) -> &'a ListLink<'a, RRProcessNode> {
+        &self.next
+    }
+}
+
+/// Round Robin Scheduler
+pub struct RoundRobinSched<'a> {
+    kernel: &'static Kernel,
+    pub processes: List<'a, RRProcessNode<'a>>,
+}
+
+impl<'a> RoundRobinSched<'a> {
+    /// How long a process can run before being pre-empted
+    const DEFAULT_TIMESLICE_US: u32 = 10000;
+    /// Skip re-scheduling a process if its quanta is nearly exhausted
+    const MIN_QUANTA_THRESHOLD_US: u32 = 500;
+    pub const fn new(kernel: &'static Kernel) -> RoundRobinSched<'a> {
+        RoundRobinSched {
+            kernel: kernel,
+            processes: List::new(),
+        }
+    }
+
+    /// Executes a process with a timeslice of DEFAULT_TIMESLICE_US -- unless the caller
+    /// indicates that this process is being rescheduled after being interrupted, in which
+    /// case the process is executed with the timeslice remaining when it was interrupted.
+    /// Returns true if the process exited because of being interrupted.
+    unsafe fn do_process<P: Platform, C: Chip>(
+        &self,
+        platform: &P,
+        chip: &C,
+        process: &dyn process::ProcessType,
+        ipc: Option<&crate::ipc::IPC>,
+        rescheduled: bool,
+    ) -> bool {
+        let systick = chip.systick();
+
+        if !rescheduled {
+            systick.reset();
+            systick.set_timer(Self::DEFAULT_TIMESLICE_US);
+        }
+        systick.enable(false); //resumes counting down
```

Comment:
The current scheduler takes the same approach of charging kernel overhead to process time quanta. In the case that a process becomes ready because of a callback, but is not yet in the Running state, this means that overhead of handling that callback is charged to the process before the process actually begins executing.

I could go either way on which approach is better.

Reply:
I'm not enough of a schedulers / real-time person to know offhand, but I imagine this decision is one that's well defined in the theory. Thinking a little more, my instinct is that to ensure schedule reasoning, this time has to be accounted for somewhere, but I could believe that the standard is to charge it to the "kernel thread" or something (or certainly it could be to the process).

Ping @phil-levis, is this something that's standardized that you know of?

Reply:
I do not have nearly the historical knowledge that Phil has, but I do know quite a bit about realtime systems, so I'll but in anyway :-)

I can only think of two situations where a thread would hit its timeslice limit in a well-designed realtime system:

1. The thread is a background thread that performs non-realtime work, such as logging. This case can be treated the same as non-realtime systems.

1. The thread is a realtime thread that overran its deadline (i.e. it is buggy). In this case, the timeslice exists to protect *other* realtime threads from the buggy realtime thread. The other threads care about total latency, which is the sum of the kernel and userspace time that preempted them, so both should count towards the timeslice limit.

Reply:
From some reading today, as far as I can tell the Linux scheduler charges in-kernel work (such as handling syscalls) to the userspace thread that made the syscall.